File: python/src/topology/is_closed.cpp
/*
 * Copyright (c) 2025 XLAB
 * All rights reserved.
 *
 * This file is part of trueform (trueform.polydera.com)
 *
 * Licensed for noncommercial use under the PolyForm Noncommercial
 * License 1.0.0.
 * Commercial licensing available via info@polydera.com.
 *
 * Author: Žiga Sajovic
 */

#include "trueform/python/topology/is_closed.hpp"
#include <nanobind/nanobind.h>
#include <nanobind/ndarray.h>

namespace tf::py {

auto register_topology_is_closed(nanobind::module_ &m) -> void {
  using namespace nanobind;

  // int32, ngon=3
  m.def(
      "is_closed_int_3",
      [](ndarray<numpy, int, shape<-1, 3>> cells,
         const offset_blocked_array_wrapper<int, int> &fm) {
        return is_closed<int, 3>(cells, fm);
      },
      arg("cells"), arg("face_membership"));

  // int32, dynamic
  m.def(
      "is_closed_int_dyn",
      [](const offset_blocked_array_wrapper<int, int> &cells,
         const offset_blocked_array_wrapper<int, int> &fm) {
        return is_closed_dynamic<int>(cells, fm);
      },
      arg("cells"), arg("face_membership"));

  // int64, ngon=3
  m.def(
      "is_closed_int64_3",
      [](ndarray<numpy, int64_t, shape<-1, 3>> cells,
         const offset_blocked_array_wrapper<int64_t, int64_t> &fm) {
        return is_closed<int64_t, 3>(cells, fm);
      },
      arg("cells"), arg("face_membership"));

  // int64, dynamic
  m.def(
      "is_closed_int64_dyn",
      [](const offset_blocked_array_wrapper<int64_t, int64_t> &cells,
         const offset_blocked_array_wrapper<int64_t, int64_t> &fm) {
        return is_closed_dynamic<int64_t>(cells, fm);
      },
      arg("cells"), arg("face_membership"));
}

} // namespace tf::py

File: python/src/trueform/_topology/is_closed.py
"""
is_closed() and is_open() function implementations

Copyright (c) 2025 Ziga Sajovic, XLAB
Licensed for noncommercial use under the PolyForm Noncommercial License 1.0.0.
Commercial licensing available via info@polydera.com.
https://github.com/polydera/trueform
"""

from .. import _trueform
from .._spatial import Mesh
from .._dispatch import topology_suffix


def is_closed(mesh: Mesh) -> bool:
    """
    Check if a mesh is closed (watertight).

    A closed mesh has no boundary edges - every edge is shared by exactly
    two faces. Closed meshes are watertight and enclose a volume.

    Parameters
    ----------
    mesh : Mesh
        The mesh to check.

    Returns
    -------
    bool
        True if the mesh is closed, False otherwise.

    Examples
    --------
    >>> import trueform as tf
    >>> import numpy as np
    >>>
    >>> # A tetrahedron is closed
    >>> faces = np.array([[0,1,2], [0,2,3], [0,3,1], [1,3,2]], dtype=np.int32)
    >>> points = np.array([[0,0,0], [1,0,0], [0.5,1,0], [0.5,0.5,1]], dtype=np.float32)
    >>> mesh = tf.Mesh(faces, points)
    >>> tf.is_closed(mesh)
    True
    >>>
    >>> # A single triangle is open
    >>> faces = np.array([[0, 1, 2]], dtype=np.int32)
    >>> points = np.array([[0,0,0], [1,0,0], [0,1,0]], dtype=np.float32)
    >>> mesh = tf.Mesh(faces, points)
    >>> tf.is_closed(mesh)
    False
    """
    if not isinstance(mesh, Mesh):
        raise TypeError(f"mesh must be Mesh, got {type(mesh).__name__}")

    if not mesh.is_dynamic and mesh.ngon != 3:
        raise ValueError(
            f"mesh must have triangular faces or be dynamic, got {mesh.ngon} vertices per face."
        )

    faces = mesh.faces
    fm = mesh._wrapper.face_membership_array()

    ngon = 'dyn' if mesh.is_dynamic else '3'
    suffix = topology_suffix(faces.dtype, ngon)
    func_name = f"is_closed_{suffix}"
    cpp_func = getattr(_trueform.topology, func_name)

    if mesh.is_dynamic:
        return cpp_func(mesh._wrapper.faces_array(), fm)
    return cpp_func(faces, fm)


def is_open(mesh: Mesh) -> bool:
    """
    Check if a mesh is open (has boundary edges).

    An open mesh has at least one boundary edge - an edge that belongs
    to only one face.

    Parameters
    ----------
    mesh : Mesh
        The mesh to check.

    Returns
    -------
    bool
        True if the mesh is open, False otherwise.

    Examples
    --------
    >>> import trueform as tf
    >>> import numpy as np
    >>>
    >>> # A single triangle is open
    >>> faces = np.array([[0, 1, 2]], dtype=np.int32)
    >>> points = np.array([[0,0,0], [1,0,0], [0,1,0]], dtype=np.float32)
    >>> mesh = tf.Mesh(faces, points)
    >>> tf.is_open(mesh)
    True
    """
    return not is_closed(mesh)
